extension/parquet/parquet-extension.cpp

```diff
@@ -303,11 +321,88 @@ class ParquetScanFunction {
 	                                                        vector<LogicalType> &return_types, vector<string> &names,
 	                                                        ParquetOptions parquet_options) {
 		auto result = make_unique<ParquetReadBindData>();
+
+		// The most likely path (Parquet Scan without union by name option)
+		if (!parquet_options.union_by_name) {
+			result->files = move(files);
+			result->SetInitialReader(make_shared<ParquetReader>(context, result->files[0], parquet_options));
+			return_types = result->types = result->initial_reader->return_types;
+			names = result->names = result->initial_reader->names;
+			return move(result);
+		} else {
+			return ParquetUnionNamesBind(context, files, return_types, names, parquet_options);
+		}
+	}
+
+	static unique_ptr<FunctionData> ParquetUnionNamesBind(ClientContext &context, vector<string> files,
+	                                                      vector<LogicalType> &return_types, vector<string> &names,
+	                                                      ParquetOptions parquet_options) {
+		auto result = make_unique<ParquetReadBindData>();
 		result->files = move(files);
 
-		result->SetInitialReader(make_shared<ParquetReader>(context, result->files[0], parquet_options));
-		return_types = result->types = result->initial_reader->return_types;
-		names = result->names = result->initial_reader->names;
+		case_insensitive_map_t<idx_t> union_names_map;
+		vector<string> union_col_names;
+		vector<LogicalType> union_col_types;
+		idx_t union_names_index = 0;
+
+		for (idx_t file_idx = 0; file_idx < result->files.size(); ++file_idx) {
+			auto reader = make_shared<ParquetReader>(context, result->files[file_idx], parquet_options);
+			auto &col_names = reader->names;
+			auto &sql_types = reader->return_types;
+			D_ASSERT(col_names.size() == sql_types.size());
+
+			// union_col_names will exclude generated columns
+			// like filename, hivepartition etc.
+			for (idx_t col = 0; col <= reader->last_parquet_col; ++col) {
```

Comment:
This code looks very similar to the code in `read_csv.cpp` - could we extract this into a separate class (e.g. `UnionByName` - similar to how we handle hive partitioning in the `HivePartitioning` class)?

Reply:
Yes! That makes code much reusable